Address containers by their names instead of ports

kremso opened the pull request:
I've implemented the switch to address containers by their assigned names rather than by their port bindings. This paves the path for having multiple containers per host and it also allows to deploy containers that run in the host networking mode (there are no port mappings in this mode).

relistan commented:
@kremso Thanks! This is a much needed change. Reviewing now.

relistan commented:
This is a great step in the right direction. With a few fixes it would replicate the current functionality without breaking anything. The biggest thing that needs to be fixed is making sure we can tell which of potentially multiple containers with the same service name we're addressing if we don't use ports. We might have to look up the container ID early on and pass that around instead.

kremso commented:
@relistan Thanks for the review and sorry for the delay. I replied to your comments.

relistan commented:
@kremso no worries! Very happy to have you taking this on. Thank *you*!

relistan commented:
Looks like with those fixes we can probably merge this.

kremso commented:
@relistan I pushed some commits with fixes. I believe I addressed all of your comments. 

For the stop phase, I retained the port lookup. I cannot figure out how to avoid using ports here without breaking backwards compatibility :(

For the start phase and health check, I changed that to a lookup by container id. It makes sense to me as it's unambiguous and I think that it cleaned up some parts of the code nicely.

relistan commented:
@kremso sorry for the delay, will review tomorrow

relistan commented:
Sorry for the delay, finally testing this out today and tried to deploy an existing project. I immediately hit this error:
```
I, [2015-05-03T10:37:01.763282 #52743]  INFO -- : Creating new container for 42a3cf88
/home/centurion/dev/centurion/lib/centurion/docker_via_api.rb:67:in `create_container': #<Excon::Response:0x007ff36b1fd4a0 @data={:body=>"Could not apply volumes of non-existent container \"parent\".\n", :headers=>{"Content-Type"=>"text/plain; charset=utf-8", "Date"=>"Sun, 03 May 2015 17:37:01 GMT", "Content-Length"=>"60"}, :status=>500, :reason_phrase=>"Internal Server Error", :remote_ip=>"172.16.168.179", :local_port=>58899, :local_address=>"172.16.168.1"}, @body="Could not apply volumes of non-existent container \"parent\".\n", @headers={"Content-Type"=>"text/plain; charset=utf-8", "Date"=>"Sun, 03 May 2015 17:37:01 GMT", "Content-Length"=>"60"}, @status=500, @remote_ip="172.16.168.179", @local_port=58899, @local_address="172.16.168.1"> (RuntimeError)
	from /home/centurion/dev/centurion/lib/centurion/deploy.rb:177:in `start_container_with_config'
	from /home/centurion/dev/centurion/lib/centurion/deploy.rb:143:in `start_new_container'
	from /home/centurion/dev/centurion/lib/tasks/deploy.rake:119:in `block (3 levels) in <top (required)>'
	from /home/centurion/dev/centurion/lib/centurion/deploy_dsl.rb:6:in `call'
	from /home/centurion/dev/centurion/lib/centurion/deploy_dsl.rb:6:in `block (2 levels) in on_each_docker_host'
	from /home/centurion/dev/centurion/lib/centurion/docker_server_group.rb:22:in `call'
	from /home/centurion/dev/centurion/lib/centurion/docker_server_group.rb:22:in `block in each'
	from /home/centurion/dev/centurion/lib/centurion/docker_server_group.rb:20:in `each'
	from /home/centurion/dev/centurion/lib/centurion/docker_server_group.rb:20:in `each'
	from /home/centurion/dev/centurion/lib/centurion/deploy_dsl.rb:6:in `block in on_each_docker_host'
	from /home/centurion/dev/centurion/lib/centurion/deploy_dsl.rb:6:in `tap'
	from /home/centurion/dev/centurion/lib/centurion/deploy_dsl.rb:6:in `on_each_docker_host'
	from /home/centurion/dev/centurion/lib/tasks/deploy.rake:116:in `block (2 levels) in <top (required)>'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:240:in `call'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:240:in `block in execute'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:235:in `each'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:235:in `execute'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:179:in `block in invoke_with_call_chain'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:172:in `invoke_with_call_chain'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:165:in `invoke'
	from /home/centurion/dev/centurion/lib/capistrano_dsl.rb:88:in `invoke'
	from /home/centurion/dev/centurion/lib/tasks/deploy.rake:22:in `block in <top (required)>'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:240:in `call'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:240:in `block in execute'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:235:in `each'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:235:in `execute'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:179:in `block in invoke_with_call_chain'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:172:in `invoke_with_call_chain'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/2.2.0/rake/task.rb:165:in `invoke'
	from /home/centurion/dev/centurion/lib/capistrano_dsl.rb:88:in `invoke'
	from /home/centurion/dev/centurion/bin/centurion:80:in `<top (required)>'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bin/centurion:23:in `load'
	from /home/centurion/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bin/centurion:23:in `<main>'
```
This is with one host on Docker 1.5.0 and a second on Docker 1.6.0. I'm looking into the issue now.

relistan commented:
Looks like the above issue is a Docker 1.6.0 breaking API change. I'll look into fixing that separately.

relistan commented:
@kremso do you want to leave this open and rebase on the other code, or close in favor of a new one?

kremso commented:
I will rebase this.

relistan commented:
Thanks!

relistan commented:
@kremso let me know if you need some help

kremso commented:
@relistan Sorry, I was busy with something else. I'll work on in this week, hopefully.

kremso commented:
@relistan Squashed and rebased on top of master.

relistan commented:
@kremso, sorry about the delay, been at Velocity. Will get this merged just after I release 1.7.0 today. Thanks!!

relistan commented:
@kremso awesome work. Merged! Thanks
